File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/remoteconfiguration/ConfigurationFetcher.java
package com.snowplowanalytics.snowplow.internal.remoteconfiguration;

import android.content.Context;
import android.net.TrafficStats;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.util.Consumer;

import com.snowplowanalytics.snowplow.configuration.RemoteConfiguration;
import com.snowplowanalytics.snowplow.internal.emitter.Executor;
import com.snowplowanalytics.snowplow.internal.tracker.Logger;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class ConfigurationFetcher {
    private final String TAG = ConfigurationFetcher.class.getSimpleName();
    private static final int TRAFFIC_STATS_TAG = 1;

    private Future<ResponseBody> future;

    @NonNull
    private final RemoteConfiguration remoteConfiguration;
    @NonNull
    private final Consumer<FetchedConfigurationBundle> onFetchCallback;

    public ConfigurationFetcher(@NonNull Context context, @NonNull RemoteConfiguration remoteConfiguration, @NonNull Consumer<FetchedConfigurationBundle> onFetchCallback) {
        this.remoteConfiguration = remoteConfiguration;
        this.onFetchCallback = onFetchCallback;
        Executor.execute(getRunnable(context), (Throwable t) -> {
            exceptionHandler(t);
        });
    }

    // Private methods

    private Runnable getRunnable(Context context) {
        return new Runnable() {
            @Override
            public void run() {
                try {
                    ResponseBody body = performRequest(context, remoteConfiguration.endpoint);
                    if (body != null) {
                        resolveRequest(context, body, onFetchCallback);
                    }
                } catch (Exception e) {
                    Logger.e(TAG, "Unable to get remote configuration: "+e.getMessage(), e);
                }
            }
        };
    }

    private ResponseBody performRequest(@NonNull Context context, @NonNull String endpoint) throws IOException {
        Uri.Builder uriBuilder = Uri.parse(endpoint).buildUpon();
        String uri = uriBuilder.build().toString();
        OkHttpClient client = new OkHttpClient.Builder()
                .connectTimeout(15, TimeUnit.SECONDS)
                .readTimeout(15, TimeUnit.SECONDS)
                .build();
        Request request = new okhttp3.Request.Builder()
                .url(uri)
                .get()
                .build();
        TrafficStats.setThreadStatsTag(TRAFFIC_STATS_TAG);
        Response resp = client.newCall(request).execute();
        ResponseBody body = resp.body();
        if (resp.isSuccessful() && body != null) {
            return body;
        }
        return null;
    }

    private void resolveRequest(@NonNull Context context, @NonNull ResponseBody responseBody, Consumer<FetchedConfigurationBundle> onFetchCallback) throws IOException, JSONException {
        String data = responseBody.string();
        JSONObject jsonObject = new JSONObject(data);
        FetchedConfigurationBundle bundle = new FetchedConfigurationBundle(context, jsonObject);
        onFetchCallback.accept(bundle);
    }

    private void exceptionHandler(@Nullable Throwable t) {
        String message = t.getMessage();
        if (message == null) {
            message = "no message provided";
        }
        Logger.e(TAG, message, t);
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/remoteconfiguration/ConfigurationFetcher.kt
package com.snowplowanalytics.core.remoteconfiguration

import android.content.Context
import android.net.TrafficStats
import android.net.Uri
import androidx.core.util.Consumer

import com.snowplowanalytics.core.emitter.Executor.execute
import com.snowplowanalytics.core.tracker.Logger
import com.snowplowanalytics.snowplow.configuration.RemoteConfiguration

import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.ResponseBody

import org.json.JSONException
import org.json.JSONObject
import java.io.IOException

import java.util.concurrent.TimeUnit

class ConfigurationFetcher(
    context: Context,
    private val remoteConfiguration: RemoteConfiguration,
    private val onFetchCallback: Consumer<FetchedConfigurationBundle>
) {
    private val TAG = ConfigurationFetcher::class.java.simpleName

    init {
        execute(getRunnable(context)) { t: Throwable? -> exceptionHandler(t) }
    }

    // Private methods
    private fun getRunnable(context: Context): Runnable {
        return Runnable {
            try {
                val body = performRequest(remoteConfiguration.endpoint)
                body?.let { resolveRequest(context, it, onFetchCallback) }
            } catch (e: Exception) {
                Logger.e(TAG, "Unable to get remote configuration: " + e.message, e)
            }
        }
    }

    @Throws(IOException::class)
    private fun performRequest(endpoint: String): ResponseBody? {
        val uriBuilder = Uri.parse(endpoint).buildUpon()
        val uri = uriBuilder.build().toString()
        val client: OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(15, TimeUnit.SECONDS)
            .build()
        val request: Request = Request.Builder()
            .url(uri)
            .get()
            .build()
        TrafficStats.setThreadStatsTag(TRAFFIC_STATS_TAG)
        val resp = client.newCall(request).execute()
        val body = resp.body
        return if (resp.isSuccessful && body != null) {
            body
        } else null
    }

    @Throws(IOException::class, JSONException::class)
    private fun resolveRequest(
        context: Context,
        responseBody: ResponseBody,
        onFetchCallback: Consumer<FetchedConfigurationBundle>
    ) {
        val data = responseBody.string()
        val jsonObject = JSONObject(data)
        val bundle = FetchedConfigurationBundle(context, jsonObject)
        onFetchCallback.accept(bundle)
    }

    private fun exceptionHandler(t: Throwable?) {
        var message = t!!.message
        if (message == null) {
            message = "no message provided"
        }
        Logger.e(TAG, message, t)
    }

    companion object {
        private const val TRAFFIC_STATS_TAG = 1
    }
}
